modules/apps/static/portal-osgi-web/portal-osgi-web-servlet-jsp-compiler/src/main/java/com/liferay/portal/osgi/web/servlet/jsp/compiler/internal/JspServlet.java

```diff
@@ -216,6 +221,40 @@ public void init(final ServletConfig servletConfig)
 
 		defaults.put(_INIT_PARAMETER_NAME_SCRATCH_DIR, sb.toString());
 
+		String symbolicName = _bundle.getSymbolicName();
+
+		BundleTracker<Bundle> bundleTracker = new BundleTracker(
+			_bundle.getBundleContext(), ~Bundle.UNINSTALLED, null) {
+
+			@Override
+			public Bundle addingBundle(Bundle bundle, BundleEvent bundleEvent) {
+				Dictionary<String, String> dictionary = bundle.getHeaders(
+					StringPool.BLANK);
+
+				String fragmentHost = dictionary.get(Constants.FRAGMENT_HOST);
+
+				if ((fragmentHost != null) &&
+					fragmentHost.equals(symbolicName)) {
+
+					Enumeration<URL> enumeration = bundle.findEntries(
+						"META-INF/resources", "*.jsp", true);
+
+					if (enumeration != null) {
+						defaults.put("hasFragment", "true");
+
+						close();
+					}
+				}
+
+				return bundle;
+			}
+
+		};
+
+		bundleTracker.open();
+
+		bundleTracker.close();
```

Comment:
hey @mtambara  Will this handle the case where initally when the JSP servlet is called, there is no attached fragments.  But then at some point later, the developer deploys a fragment.  Will the bundle tracker that sets the `hasFragment` be re-called?

Reply:
@gamerson Yes, when a fragment is attached the host bundle needs to be refreshed, creating a new JSP servlet and rechecking for fragment bundles

Reply:
Gotcha, makes sense.